// @azure/ai-form-recognizer/dist/commonjs/bin/defaultFields.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultResultFields = void 0;
exports.getFeatures = getFeatures;
/**
 * @internal
 * Field information for the top level fields of `AnalyzeResult`.
 */
exports.defaultResultFields = {
    pages: {
        name: "pages",
        docContents: "Extracted pages.",
        type: "fr.DocumentPage[]",
        optional: true,
    },
    tables: {
        name: "tables",
        docContents: "Extracted tables.",
        type: "fr.DocumentTable[]",
        optional: true,
    },
    keyValuePairs: {
        name: "keyValuePairs",
        docContents: "Extracted key-value pairs.",
        type: "fr.DocumentKeyValuePair[]",
        optional: true,
    },
    languages: {
        name: "languages",
        docContents: "Extracted text languages.",
        type: "fr.DocumentLanguage[]",
        optional: true,
    },
    styles: {
        name: "styles",
        docContents: "Extracted font styles.",
        type: "fr.DocumentStyle[]",
        optional: true,
    },
    unconstrainedDocuments: {
        name: "documents",
        docContents: "Extracted documents containing any extracted fields.",
        type: "fr.AnalyzedDocument[]",
        optional: true,
    },
    paragraphs: {
        name: "paragraphs",
        docContents: "Extracted document paragraphs.",
        type: "fr.DocumentParagraph[]",
        optional: true,
    },
};
const allFeatures = [
    "pages",
    "tables",
    "keyValuePairs",
    "languages",
    "styles",
    "paragraphs",
    "_documents",
];
const textFeatures = ["pages", "paragraphs", "styles"];
const layoutFeatures = [...textFeatures, "tables"];
const documentFeatures = [...layoutFeatures, "keyValuePairs"];
/**
 * @internal
 * @param model - the model to get the features of
 * @returns the list of features supported by the model
 */
function getFeatures(model) {
    var _a, _b;
    return ((_b = (_a = model.features) !== null && _a !== void 0 ? _a : {
        "prebuilt-read": [...textFeatures, "languages"],
        "prebuilt-layout": layoutFeatures,
        "prebuilt-document": [...documentFeatures, "unconstrainedDocuments"],
        "prebuilt-receipt": [...textFeatures, "keyValuePairs", "_documents"],
        "prebuilt-invoice": [...layoutFeatures, "keyValuePairs", "_documents"],
        "prebuilt-idDocument": [...textFeatures, "keyValuePairs", "_documents"],
        "prebuilt-businessCard": [...textFeatures, "keyValuePairs", "_documents"],
        "prebuilt-tax.us.w2": [...textFeatures, "keyValuePairs", "_documents"],
        "prebuilt-vaccinationCard": [...textFeatures, "keyValuePairs", "_documents"],
        "prebuilt-healthInsuranceCard.us": [...textFeatures, "keyValuePairs", "_documents"],
    }[model.modelId]) !== null && _b !== void 0 ? _b : (model.modelId.startsWith("prebuilt-") ? allFeatures : [...documentFeatures, "_documents"]));
}
