// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/ModelInfoSupportedParameterDefault.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelInfoSupportedParameterDefaultToJSON = exports.ModelInfoSupportedParameterDefaultFromJSONTyped = exports.ModelInfoSupportedParameterDefaultFromJSON = exports.instanceOfModelInfoSupportedParameterDefault = void 0;
/**
 * Check if a given object implements the ModelInfoSupportedParameterDefault interface.
 */
function instanceOfModelInfoSupportedParameterDefault(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfModelInfoSupportedParameterDefault = instanceOfModelInfoSupportedParameterDefault;
function ModelInfoSupportedParameterDefaultFromJSON(json) {
    return ModelInfoSupportedParameterDefaultFromJSONTyped(json, false);
}
exports.ModelInfoSupportedParameterDefaultFromJSON = ModelInfoSupportedParameterDefaultFromJSON;
function ModelInfoSupportedParameterDefaultFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.ModelInfoSupportedParameterDefaultFromJSONTyped = ModelInfoSupportedParameterDefaultFromJSONTyped;
function ModelInfoSupportedParameterDefaultToJSON(value) {
    return value;
}
exports.ModelInfoSupportedParameterDefaultToJSON = ModelInfoSupportedParameterDefaultToJSON;
