// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/ModelInfoMetric.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelInfoMetricToJSON = exports.ModelInfoMetricFromJSONTyped = exports.ModelInfoMetricFromJSON = exports.ModelInfoMetric = void 0;
/**
 * A distance metric that the embedding model supports for similarity searches.
 * @export
 */
exports.ModelInfoMetric = {
    Cosine: 'cosine',
    Euclidean: 'euclidean',
    Dotproduct: 'dotproduct'
};
function ModelInfoMetricFromJSON(json) {
    return ModelInfoMetricFromJSONTyped(json, false);
}
exports.ModelInfoMetricFromJSON = ModelInfoMetricFromJSON;
function ModelInfoMetricFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.ModelInfoMetricFromJSONTyped = ModelInfoMetricFromJSONTyped;
function ModelInfoMetricToJSON(value) {
    return value;
}
exports.ModelInfoMetricToJSON = ModelInfoMetricToJSON;
