// @azure/ai-form-recognizer/dist/commonjs/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.FORM_RECOGNIZER_API_VERSION = exports.SDK_VERSION = exports.DEFAULT_COGNITIVE_SCOPE = exports.KnownFormRecognizerAudience = void 0;
/**
 * Defines the known cloud audiences for Form Recognizer.
 *
 * To authenticate with Entra Id (using a `TokenCredential`) in a [Sovereign Cloud](https://learn.microsoft.com/entra/identity-platform/authentication-national-cloud)
 * environment, provide the appropriate value below as the `audience` option when creating a
 * `DocumentAnalysisClient` or `DocumentModelAdministrationClient`.
 *
 * The default value is suitable for Form Recognizer resources created in the Azure Public Cloud, so this value
 * is only required to use Form Recognizer in a different cloud environment.
 */
var KnownFormRecognizerAudience;
(function (KnownFormRecognizerAudience) {
    /** Azure China */
    KnownFormRecognizerAudience["AzureChina"] = "https://cognitiveservices.azure.cn";
    /** Azure Government */
    KnownFormRecognizerAudience["AzureGovernment"] = "https://cognitiveservices.azure.us";
    /** Azure Public Cloud */
    KnownFormRecognizerAudience["AzurePublicCloud"] = "https://cognitiveservices.azure.com";
})(KnownFormRecognizerAudience || (exports.KnownFormRecognizerAudience = KnownFormRecognizerAudience = {}));
/**
 * The default Entra Id permissions scope for Cognitive Services.
 * @internal
 */
exports.DEFAULT_COGNITIVE_SCOPE = `${KnownFormRecognizerAudience.AzurePublicCloud}/.default`;
/**
 * @internal
 */
exports.SDK_VERSION = "5.1.0";
exports.FORM_RECOGNIZER_API_VERSION = "2023-07-31";
