// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/BackupList.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupListToJSON = exports.BackupListFromJSONTyped = exports.BackupListFromJSON = exports.instanceOfBackupList = void 0;
const runtime_1 = require("../runtime");
const BackupModel_1 = require("./BackupModel");
const PaginationResponse_1 = require("./PaginationResponse");
/**
 * Check if a given object implements the BackupList interface.
 */
function instanceOfBackupList(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfBackupList = instanceOfBackupList;
function BackupListFromJSON(json) {
    return BackupListFromJSONTyped(json, false);
}
exports.BackupListFromJSON = BackupListFromJSON;
function BackupListFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'data': !(0, runtime_1.exists)(json, 'data') ? undefined : (json['data'].map(BackupModel_1.BackupModelFromJSON)),
        'pagination': !(0, runtime_1.exists)(json, 'pagination') ? undefined : (0, PaginationResponse_1.PaginationResponseFromJSON)(json['pagination']),
    };
}
exports.BackupListFromJSONTyped = BackupListFromJSONTyped;
function BackupListToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'data': value.data === undefined ? undefined : (value.data.map(BackupModel_1.BackupModelToJSON)),
        'pagination': (0, PaginationResponse_1.PaginationResponseToJSON)(value.pagination),
    };
}
exports.BackupListToJSON = BackupListToJSON;
