// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/CreateBackupRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateBackupRequestToJSON = exports.CreateBackupRequestFromJSONTyped = exports.CreateBackupRequestFromJSON = exports.instanceOfCreateBackupRequest = void 0;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the CreateBackupRequest interface.
 */
function instanceOfCreateBackupRequest(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfCreateBackupRequest = instanceOfCreateBackupRequest;
function CreateBackupRequestFromJSON(json) {
    return CreateBackupRequestFromJSONTyped(json, false);
}
exports.CreateBackupRequestFromJSON = CreateBackupRequestFromJSON;
function CreateBackupRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': !(0, runtime_1.exists)(json, 'name') ? undefined : json['name'],
        'description': !(0, runtime_1.exists)(json, 'description') ? undefined : json['description'],
    };
}
exports.CreateBackupRequestFromJSONTyped = CreateBackupRequestFromJSONTyped;
function CreateBackupRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'description': value.description,
    };
}
exports.CreateBackupRequestToJSON = CreateBackupRequestToJSON;
