// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/PaginationResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaginationResponseToJSON = exports.PaginationResponseFromJSONTyped = exports.PaginationResponseFromJSON = exports.instanceOfPaginationResponse = void 0;
/**
 * Check if a given object implements the PaginationResponse interface.
 */
function instanceOfPaginationResponse(value) {
    let isInstance = true;
    isInstance = isInstance && "next" in value;
    return isInstance;
}
exports.instanceOfPaginationResponse = instanceOfPaginationResponse;
function PaginationResponseFromJSON(json) {
    return PaginationResponseFromJSONTyped(json, false);
}
exports.PaginationResponseFromJSON = PaginationResponseFromJSON;
function PaginationResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'next': json['next'],
    };
}
exports.PaginationResponseFromJSONTyped = PaginationResponseFromJSONTyped;
function PaginationResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'next': value.next,
    };
}
exports.PaginationResponseToJSON = PaginationResponseToJSON;
