// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/RestoreJobList.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestoreJobListToJSON = exports.RestoreJobListFromJSONTyped = exports.RestoreJobListFromJSON = exports.instanceOfRestoreJobList = void 0;
const runtime_1 = require("../runtime");
const PaginationResponse_1 = require("./PaginationResponse");
const RestoreJobModel_1 = require("./RestoreJobModel");
/**
 * Check if a given object implements the RestoreJobList interface.
 */
function instanceOfRestoreJobList(value) {
    let isInstance = true;
    isInstance = isInstance && "data" in value;
    return isInstance;
}
exports.instanceOfRestoreJobList = instanceOfRestoreJobList;
function RestoreJobListFromJSON(json) {
    return RestoreJobListFromJSONTyped(json, false);
}
exports.RestoreJobListFromJSON = RestoreJobListFromJSON;
function RestoreJobListFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'data': (json['data'].map(RestoreJobModel_1.RestoreJobModelFromJSON)),
        'pagination': !(0, runtime_1.exists)(json, 'pagination') ? undefined : (0, PaginationResponse_1.PaginationResponseFromJSON)(json['pagination']),
    };
}
exports.RestoreJobListFromJSONTyped = RestoreJobListFromJSONTyped;
function RestoreJobListToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'data': (value.data.map(RestoreJobModel_1.RestoreJobModelToJSON)),
        'pagination': (0, PaginationResponse_1.PaginationResponseToJSON)(value.pagination),
    };
}
exports.RestoreJobListToJSON = RestoreJobListToJSON;
