// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/RestoreJobModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestoreJobModelToJSON = exports.RestoreJobModelFromJSONTyped = exports.RestoreJobModelFromJSON = exports.instanceOfRestoreJobModel = void 0;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the RestoreJobModel interface.
 */
function instanceOfRestoreJobModel(value) {
    let isInstance = true;
    isInstance = isInstance && "restoreJobId" in value;
    isInstance = isInstance && "backupId" in value;
    isInstance = isInstance && "targetIndexName" in value;
    isInstance = isInstance && "targetIndexId" in value;
    isInstance = isInstance && "status" in value;
    isInstance = isInstance && "createdAt" in value;
    return isInstance;
}
exports.instanceOfRestoreJobModel = instanceOfRestoreJobModel;
function RestoreJobModelFromJSON(json) {
    return RestoreJobModelFromJSONTyped(json, false);
}
exports.RestoreJobModelFromJSON = RestoreJobModelFromJSON;
function RestoreJobModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'restoreJobId': json['restore_job_id'],
        'backupId': json['backup_id'],
        'targetIndexName': json['target_index_name'],
        'targetIndexId': json['target_index_id'],
        'status': json['status'],
        'createdAt': (new Date(json['created_at'])),
        'completedAt': !(0, runtime_1.exists)(json, 'completed_at') ? undefined : (new Date(json['completed_at'])),
        'percentComplete': !(0, runtime_1.exists)(json, 'percent_complete') ? undefined : json['percent_complete'],
    };
}
exports.RestoreJobModelFromJSONTyped = RestoreJobModelFromJSONTyped;
function RestoreJobModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'restore_job_id': value.restoreJobId,
        'backup_id': value.backupId,
        'target_index_name': value.targetIndexName,
        'target_index_id': value.targetIndexId,
        'status': value.status,
        'created_at': (value.createdAt.toISOString()),
        'completed_at': value.completedAt === undefined ? undefined : (value.completedAt.toISOString()),
        'percent_complete': value.percentComplete,
    };
}
exports.RestoreJobModelToJSON = RestoreJobModelToJSON;
