// @azure/ai-form-recognizer/dist/commonjs/options/AnalyzeDocumentOptions.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormRecognizerFeature = void 0;
/**
 * Known feature flags supported by the Form Recognizer clients.
 */
// eslint-disable-next-line @typescript-eslint/no-redeclare
exports.FormRecognizerFeature = {
    /**
     * Enables extracting extra font information.
     */
    Fonts: "styleFont",
    /**
     * Enables high-resolution processing for documents with small text.
     */
    OcrHighResolution: "ocrHighResolution",
    /**
     * Enables the detection of mathematical expressions in the document..
     */
    Formulas: "formulas",
    /**
     * Enables the detection of the text content language.
     */
    Languages: "languages",
    /**
     *  Enables the detection of barcodes in the document.
     */
    Barcodes: "barcodes",
    /**
     *  Enables the detection of general key value pairs (form fields) in the document.
     */
    KeyValuePairs: "keyValuePairs",
};
