// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/ListNamespacesResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListNamespacesResponseToJSON = exports.ListNamespacesResponseFromJSONTyped = exports.ListNamespacesResponseFromJSON = exports.instanceOfListNamespacesResponse = void 0;
const runtime_1 = require("../runtime");
const NamespaceDescription_1 = require("./NamespaceDescription");
const Pagination_1 = require("./Pagination");
/**
 * Check if a given object implements the ListNamespacesResponse interface.
 */
function instanceOfListNamespacesResponse(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfListNamespacesResponse = instanceOfListNamespacesResponse;
function ListNamespacesResponseFromJSON(json) {
    return ListNamespacesResponseFromJSONTyped(json, false);
}
exports.ListNamespacesResponseFromJSON = ListNamespacesResponseFromJSON;
function ListNamespacesResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'namespaces': !(0, runtime_1.exists)(json, 'namespaces') ? undefined : (json['namespaces'].map(NamespaceDescription_1.NamespaceDescriptionFromJSON)),
        'pagination': !(0, runtime_1.exists)(json, 'pagination') ? undefined : (0, Pagination_1.PaginationFromJSON)(json['pagination']),
    };
}
exports.ListNamespacesResponseFromJSONTyped = ListNamespacesResponseFromJSONTyped;
function ListNamespacesResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'namespaces': value.namespaces === undefined ? undefined : (value.namespaces.map(NamespaceDescription_1.NamespaceDescriptionToJSON)),
        'pagination': (0, Pagination_1.PaginationToJSON)(value.pagination),
    };
}
exports.ListNamespacesResponseToJSON = ListNamespacesResponseToJSON;
