// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/NamespaceDescription.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-04
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamespaceDescriptionToJSON = exports.NamespaceDescriptionFromJSONTyped = exports.NamespaceDescriptionFromJSON = exports.instanceOfNamespaceDescription = void 0;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the NamespaceDescription interface.
 */
function instanceOfNamespaceDescription(value) {
    let isInstance = true;
    return isInstance;
}
exports.instanceOfNamespaceDescription = instanceOfNamespaceDescription;
function NamespaceDescriptionFromJSON(json) {
    return NamespaceDescriptionFromJSONTyped(json, false);
}
exports.NamespaceDescriptionFromJSON = NamespaceDescriptionFromJSON;
function NamespaceDescriptionFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': !(0, runtime_1.exists)(json, 'name') ? undefined : json['name'],
        'recordCount': !(0, runtime_1.exists)(json, 'record_count') ? undefined : json['record_count'],
    };
}
exports.NamespaceDescriptionFromJSONTyped = NamespaceDescriptionFromJSONTyped;
function NamespaceDescriptionToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'record_count': value.recordCount,
    };
}
exports.NamespaceDescriptionToJSON = NamespaceDescriptionToJSON;
