// @azure/ai-form-recognizer/dist/react-native/lro/administration.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { FormRecognizerError } from "../error.js";
/**
 * Convert an operation result into a training poller state.
 * @internal
 */
export async function toTrainingPollOperationState(response) {
    var _a;
    return {
        operationId: response.operationId,
        status: response.status,
        apiVersion: response.apiVersion,
        percentCompleted: (_a = response.percentCompleted) !== null && _a !== void 0 ? _a : 0,
        lastUpdatedOn: response.lastUpdatedOn,
        createdOn: response.createdOn,
        error: response.error && new FormRecognizerError(response.error),
        isCancelled: response.status === "canceled",
        isCompleted: response.status === "succeeded",
        isStarted: response.status !== "notStarted",
        tags: response.tags,
        // The following assertion is required. Technically the type of `response.result` is
        // `DocumentModelDetails | DocumentClassifierDetails | undefined`, which isn't assignable to the type of
        // either operation state's result. We would need some kind of dependent typing to express how the type of `result`
        // actually _determines_ the type of the resulting return value.
        result: response.result,
    };
}
